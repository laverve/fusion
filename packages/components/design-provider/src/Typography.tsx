import React, { JSXElementConstructor } from "react";

type TypographyComponent = React.PropsWithChildren & { className: string; style?: React.CSSProperties };
type TypographyVariant = "heading-1" | "heading-2" | "heading-3" | "heading-4" | "body-1" | "body-2" | "body-3";
export type TypographyProps = React.PropsWithChildren & {
    Component?: string | JSXElementConstructor<TypographyComponent>;
    variant?: TypographyVariant;
    className?: string;
    style?: React.CSSProperties;
};

const VARIANT_COMPONENT_MAPPING = new Map<TypographyVariant, string | JSXElementConstructor<TypographyComponent>>([
    ["heading-1", "h1"],
    ["heading-2", "h2"],
    ["heading-3", "h3"],
    ["heading-4", "h4"],
    ["body-1", "p"],
    ["body-2", "p"],
    ["body-3", "p"]
]);

export const Typography: React.FC<TypographyProps> = ({
    Component,
    variant = "body-1",
    className = "",
    children,
    style
}) => {
    const InternalComponent: string | JSXElementConstructor<TypographyComponent> =
        Component || VARIANT_COMPONENT_MAPPING.get(variant) || "div";
    return (
        <InternalComponent className={`${variant} ${className}`} style={style}>
            {children}
        </InternalComponent>
    );
};
