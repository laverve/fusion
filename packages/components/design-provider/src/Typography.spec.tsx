import { describe, it, expect } from "@jest/globals";
import React from "react";
import renderer from "react-test-renderer";
import { Typography, TypographyProps } from "./Typography";

describe("Typography", () => {
    it("renders with default props", () => {
        const component = renderer.create(<Typography>Test</Typography>);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("supports custom component", () => {
        const component = renderer.create(<Typography Component="div">Test</Typography>);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("supports a set of variants", () => {
        const variants: TypographyProps["variant"][] = [
            "heading-1",
            "heading-2",
            "heading-3",
            "heading-4",
            "body-1",
            "body-2",
            "body-3"
        ];
        variants.forEach((variant) => {
            const component = renderer.create(<Typography variant={variant}>Test</Typography>);
            expect(component.toJSON()).toMatchSnapshot();
        });
    });

    it("supports custom className", () => {
        const component = renderer.create(<Typography className="test-class-name">Test</Typography>);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("supports custom styles", () => {
        const component = renderer.create(<Typography style={{ textDecoration: "underline" }}>Test</Typography>);
        expect(component.toJSON()).toMatchSnapshot();
    });
});
