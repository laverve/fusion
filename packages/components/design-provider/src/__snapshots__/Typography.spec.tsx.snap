// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography renders with default props 1`] = `
<p
  className="body-1 "
>
  Test
</p>
`;

exports[`Typography supports a set of variants 1`] = `
<h1
  className="heading-1 "
>
  Test
</h1>
`;

exports[`Typography supports a set of variants 2`] = `
<h2
  className="heading-2 "
>
  Test
</h2>
`;

exports[`Typography supports a set of variants 3`] = `
<h3
  className="heading-3 "
>
  Test
</h3>
`;

exports[`Typography supports a set of variants 4`] = `
<h4
  className="heading-4 "
>
  Test
</h4>
`;

exports[`Typography supports a set of variants 5`] = `
<p
  className="body-1 "
>
  Test
</p>
`;

exports[`Typography supports a set of variants 6`] = `
<p
  className="body-2 "
>
  Test
</p>
`;

exports[`Typography supports a set of variants 7`] = `
<p
  className="body-3 "
>
  Test
</p>
`;

exports[`Typography supports custom className 1`] = `
<p
  className="body-1 test-class-name"
>
  Test
</p>
`;

exports[`Typography supports custom component 1`] = `
<div
  className="body-1 "
>
  Test
</div>
`;

exports[`Typography supports custom styles 1`] = `
<p
  className="body-1 "
  style={
    {
      "textDecoration": "underline",
    }
  }
>
  Test
</p>
`;
